### ISSUE TITLE
Can't set password while using ec2-discovery with ec2:privateIpv4 interface

### ISSUE THREAD

**Issue by jwsomis** [elastic/elasticsearch#33580]
**Elasticsearch version** (`bin/elasticsearch --version`): 6.3.1

**Plugins installed**: [discovery-ec2, repository-s3]

**JVM version** (`java -version`): 1.8.0_181

**OS version** (`uname -a` if on a Unix-like system): Amazon Linux 2

**Description of the problem including expected versus actual behavior**:
I have discovery-ec2 installed and working with the below config
```
network.host: ["_ec2:privateIpv4_", "_local_"]
discovery.zen:
  hosts_provider: ec2
discovery.ec2:
  endpoint: https://ec2.<region>.amazonaws.com
  tag.asg_name: <tag>
```
If I try to then set passwords, I get "No interface named 'ec2:privateIpv4' found"

**Steps to reproduce**:

1. Configure an elasticsearch cluster using discovery-ec2
2. Run /usr/share/elasticsearch/bin/elasticsearch-setup-passwords interactive
    2a. Same issue with auto
 
**Provide logs (if relevant)**:
```
Exception in thread "main" java.lang.IllegalArgumentException: No interface named 'ec2:privateIpv4' found, got [name:lo (lo), name:eth0 (eth0)]
	at org.elasticsearch.common.network.NetworkUtils.getAddressesForInterface(NetworkUtils.java:231)
	at org.elasticsearch.common.network.NetworkService.resolveInternal(NetworkService.java:233)
	at org.elasticsearch.common.network.NetworkService.resolveInetAddresses(NetworkService.java:187)
	at org.elasticsearch.common.network.NetworkService.resolvePublishHostAddresses(NetworkService.java:145)
	at org.elasticsearch.xpack.security.authc.esnative.tool.CommandLineHttpClient.getDefaultURL(CommandLineHttpClient.java:147)
	at org.elasticsearch.xpack.security.authc.esnative.tool.SetupPasswordTool$SetupCommand.setupOptions(SetupPasswordTool.java:266)
	at org.elasticsearch.xpack.security.authc.esnative.tool.SetupPasswordTool$InteractiveSetup.execute(SetupPasswordTool.java:173)
	at org.elasticsearch.cli.EnvironmentAwareCommand.execute(EnvironmentAwareCommand.java:86)
	at org.elasticsearch.cli.Command.mainWithoutErrorHandling(Command.java:124)
	at org.elasticsearch.cli.MultiCommand.execute(MultiCommand.java:79)
	at org.elasticsearch.cli.Command.mainWithoutErrorHandling(Command.java:124)
	at org.elasticsearch.cli.Command.main(Command.java:90)
	at org.elasticsearch.xpack.security.authc.esnative.tool.SetupPasswordTool.main(SetupPasswordTool.java:106)
```

**Comment by ywelsch** [comment]
@jaymode can you comment on this? (I see that you've worked on this and added a comment that using custom providers is not possible?)

**Comment by albertzaharovits** [comment]
I think it requires to create a `Node` instance to be able to properly resolve custom network addresses
and this might not be possible given that ports are bound (and maybe other resource clashes) and in general the heavy handed approach of booting up a node inside a command line tool is not desirable.

I think the command line should parse the [ports file](https://github.com/elastic/elasticsearch/blob/a69ae6b89f0560b2dfc97e86eebd1268e945e2b9/server/src/main/java/org/elasticsearch/node/Node.java#L770) and, on the server side, we should default to writing the ports file. The command line tool should bail if the ports file is not written and the `--url` option to the cmd line is not specified.

I have opened: https://github.com/elastic/elasticsearch/issues/33666

**Comment by radiating** [comment]
Hi, is there a work around for this issue yet? Thanks!

**Comment by tvernum** [comment]
@radiating Just use the `--url` parameter to set the URL for your Elasticsearch node.

The tool tries to guess you local ES URL by reading the config file, but that isn't possible in all cases, so if you are using a non-default discovery provider, then you may need to explicitly pass in the URL.

**Comment by henningandersen** [comment]
@radiating and @jwsomis were you able to proceed by using the `--url` parameter as described by @tvernum ? If so, can we close this issue?

**Comment by radiating** [comment]
@henningandersen yes it worked!

**Comment by ywelsch** [comment]
I think we should treat the auto-detection logic as best-effort and I've opened #40151 to that effect, enhancing the exception message to make it clear what needs to be done.

**Comment by ywelsch** [comment]
Closed by #40151, which treats the URL auto-detection logic as best effort and improves the exception message to make it clearer that an URL needs to be explicitly passed in case it cannot be auto-deduced.
